src/context.py

```diff
@@ -0,0 +1,69 @@
+Features = {"CASEMAPPING": "rfc1459", "CHARSET": "utf-8", "STATUSMSG": {"@", "+"}, "CHANTYPES": {"#"}} # IRC server features (these are defaults)
+
+def lower(nick):
+    if nick is None:
+        return None
+    if isinstance(nick, IRCContext):
+        return nick.lower()
+
+    mapping = {
+        "[": "{",
+        "]": "}",
+        "\\": "|",
+        "^": "~",
+    }
+
+    if Features["CASEMAPPING"] == "strict-rfc1459":
+        mapping.pop("^")
+    elif Features["CASEMAPPING"] == "ascii":
+        mapping.clear()
+
+    return nick.lower().translate(str.maketrans(mapping))
+
+class IRCContext:
+    """Base class for channels and users."""
+
+    is_channel = False
+    is_user = False
+    is_fake = False
+
+    def __init__(self, name, client, *, ref=None):
+        self.name = name
+        self.client = client
+        self.ref = ref
+
+    def lower(self):
+        return type(self)(lower(name), client, ref=(self.ref or ref))
+
+    def get_send_type(self, *, is_notice=False, is_privmsg=False):
+        if is_notice and not is_privmsg:
+            return "NOTICE"
+        return "PRIVMSG"
+
+    @staticmethod
+    def raw_send(data, client, send_type, name):
+        full_address = "{cli.nickname}!{cli.ident}@{cli.hostmask}".format(cli=client)
+
+        # Maximum length of sent data is 0x200 (512) bytes. However,
+        # we have to reduce the maximum length allowed to account for:
+        # 1 (1) - The initial colon at the front of the data
+        # 2 (1) - The space between the command and the target
+        # 2 (1) - The space between the target and the data
+        # 3 (1) - The colon at the front of the data to send
+        # 4 (3) - I don't know why, but we need 3 more/less characters
+        length = 0x200 - 7
+        # Next, we need to reduce the length to account for our address
+        length -= len(full_address)
+        # Then we also need to account for the target's length
+        length -= len(name)
+        # Finally, we need to account for the send type's length
+        length -= len(send_type)
```

Comment:
This is incredibly flawed. You aren't including full_address in the sent message so there is no need to deduct it. Remember also that the trailing \r\n is included in the 512 byte count. Don't be clever with 0x200, it's less obvious than 512 -- just use that.

Your comment has redundant lines and incorrect lines, clean it up